### PR TITLE
Love: Add optional title to hui-glance-card

### PR DESCRIPTION
**Pull request by jeradM**
Allows for an optional title for each entity in the glance card. This way a shorter name can be displayed to compensate for the reduced horizontal space.

This is breaking:

```yaml
- type: glance
  title: Living Room
  entities:
    - entity_id: sensor.living_room_temperature
      title: Temp
    - entity_id: sensor.living_room_luminance
      title: Light

### CODE REVIEW

**Comment by c727** [comment]
@balloob do we want to drop the short syntax for entities (for all cards) or shall all cards support both?

also use `entity` instead of `entity_id`

**Comment by balloob** [comment]
I think we can keep both. What do you think?

File: src/panels/lovelace/cards/hui-glance-card.js [hunk 1]
```diff
@@ -105,7 +105,8 @@ class HuiGlanceCard extends LocalizeMixin(EventsMixin(PolymerElement)) {
   }
 
   _computeEntities(config) {
-    if (config && config.entities && Array.isArray(config.entities)) {
+    if (config && config.entities && Array.isArray(config.entities) &&
+        config.entities.every(x => Object.prototype.hasOwnProperty.call(x, 'entity_id'))) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-glance-card.js]: I prefer the synax 'entity' in x

**Comment by c727** [comment]
for us it's easier if we only have one syntax but users won't like it. Idk

**Comment by jeradM** [comment]
I'll change `entity_id` to `entity`.  Are you guys talking about allowing both the old simple entity syntax and this new syntax together?

```yaml
- type: glance
  title: Living Room
  entities:
    - entity: sensor.living_room_temperature
      title: Temp
    - sensor.living_room_luminance
```
If so, I can add that to this PR. Should just be as simple as checking if `item` is a string and creating an object out of it

**Comment by jeradM** [comment]
Should this maybe be in a helper function somewhere that takes `entities`, turns the plain entity_ids into `{entity: 'entity_id'}`, and returns an array of objects? Would make it cleaner if other cards need this type of functionality

**Comment by jeradM** [comment]
Cool, I'll work on this. Is `src/common/entity/` the right place for this helper?

I'm happy to help where I can 😄, lovelace is exciting and you're doing awesome work on it

**Comment by c727** [comment]
we keep lovelace stuff in `/src/panels/lovelace/common/` or `/src/panels/lovelace/components/`

**Comment by c727** [comment]
what do you think about this?
```javascript
export default function computeConfigEntities(config) {
  if (!config || !config.entities) {
    return null;
  }

  const entities = config.entities;
  if (typeof entities === 'string') {
    return [{ entity: entities }];
  }

  if (Array.isArray(entities)) {
    return entities.map((entity) => {
      if (typeof entity === 'string') {
        return { entity };
      } else if (typeof entity === 'object' && !Array.isArray(entity)) {
        return entity;
      }
      return null;
    });
  }
  return null;
}
```

File: src/panels/lovelace/common/coerce-entities-to-objects.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/** Coerce plain entity_id strings into {entity: 'entity_id'} objects **/
+export default function coerceEntitiesToObjects(entities) {
+  if (!Array.isArray(entities))
```
Comment (balloob) [inline src/panels/lovelace/common/coerce-entities-to-objects.js]: Don't do this, raise instead. Fail loudly.

File: src/panels/lovelace/common/coerce-entities-to-objects.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    return entities;
+
+  const result = [];
+  entities.forEach(entity => {
+    if (typeof entity === 'string' || entity instanceof String) {
```
Comment (balloob) [inline src/panels/lovelace/common/coerce-entities-to-objects.js]: This check is the same check twice?

Comment (c727) [reply src/panels/lovelace/common/coerce-entities-to-objects.js]: the 2nd is for `x = new String("test")` but Imo we don't need this

File: src/panels/lovelace/common/coerce-entities-to-objects.js (hunk 2 above)
Comment (jeradM) [reply src/panels/lovelace/common/coerce-entities-to-objects.js]: That's weird, that isn't the most recent commit. I'm not sure why you're still seeing that version. Anyway, this isn't necessarily a failure. I tried to make this as general as possible so in the most recent commit, if `entities` is not an array it still works, it just does the conversion and returns a single entity, but I am liking @c727 version better; just take the config and return the entities

File: src/panels/lovelace/common/coerce-entities-to-objects.js (hunk 3 above)
Comment (jeradM) [reply src/panels/lovelace/common/coerce-entities-to-objects.js]: Yeah, it's probably overkill here. I'll remove it

**Comment by jeradM** [comment]
@c727 I do like your approach better; just take the config and return entities, the only issue I see is that it doesn't handle single entities (e.g. the entity-picture card). Maybe we should check for `config.entity` and if it exists, assume a single entity config, otherwise check for `config.entities`?

**Comment by c727** [comment]
I think for `config.entity` we should create a separate file/function once we need it. For this card we use `config.entities`

**Comment by jeradM** [comment]
@c727 One more option before I update. How about checking that all entities are valid in `computeConfigEntities()`?

```javascript
export default function computeConfigEntities(config) {
  let entities = config && config.entities;

  if (entities && Array.isArray(entities) {
    entities = entities.map((entity) => {
      if (typeof entity === 'string') {
        return { entity };
      }
      return entity;
    });
  
    if (entities.every(x => typeof x === 'object' && !Array.isArray(x) && 'entity' in x)) {
      return entities;
    }
  }

  throw Error("Error in card configuration");
}
```
This will force `entities` to be a declared as a list and fail hard if config is invalid. Then in the card, `_computeEntities()` could just do
```javascript
this._error = null;
try {
  return computeConfigEntities(config);
} catch (e) {
  this._error = e.message;
  return [];
}
```

**Comment by c727** [comment]
I would return null on error. Aslo do validation first and return on error as shown above. You can use entities.some(x === null) to check if all have x.entity

**Comment by jeradM** [comment]
Add the `'entity' in entity` check to the `else`? 
```javascript
export default function computeConfigEntities(config) {
  const entities = config && config.entities;

  if (!entities) {
    return null;
  }

  if (typeof entities === 'string') {
    return [{ entity: entities }];
  }

  if (Array.isArray(entities)) {
    return entities.map((entity) => {
      if (typeof entity === 'string') {
        return { entity };
      } else if (typeof entity === 'object' && !Array.isArray(entity) && 'entity' in entity) {
        return entity;
      }
      return null;
    });
  }

  return null;
}
```

**Comment by c727** [comment]
you can also join our discord chat again, balloob made some notes related to config validation in frontend dev channel

**Review by balloob — APPROVED** [review]

**Comment by balloob** [comment]
ok to merge once conflict resolved.

**Comment by jeradM** [comment]
@c727 beat me to it. Can I still push one more commit that reverts back to using `computeConfigEntities` and `validateEntitiesConfig`?

**Pull request merged**